modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseUpdater.java

```diff
@@ -443,8 +443,19 @@ private void updateLeaseBatchInternal() {
                             ? lease.getLeaseholder()
                             : lease.proposedCandidate();
 
-                    ClusterNode candidate = nextLeaseHolder(assignments, grpId, proposedLeaseholder);
+                    ClusterNode candidate = nextLeaseHolder(stableAssignments, grpId, proposedLeaseholder);
 
+                    // if there wasn't a candidate among stable assignments set then make attempt to select a candidate among pending set
+                    if (candidate == null) {
+                        Map<ReplicationGroupId, TokenizedAssignments> pendingMap = assignmentsTracker.pendingAssignments();
+
+                        if (pendingMap.containsKey(grpId)) {
+                            Set<Assignment> pendingAssignments = pendingMap.get(grpId).nodes();
+                            candidate = nextLeaseHolder(pendingAssignments, grpId, proposedLeaseholder);
+                        }
+                    }
```

Comment:
I doubt there, because in the first scenario from the ticket lease on B is trying to expire (time is out), but we can't to find a proper candidate in this lines. So, the case `lease.getExpirationTime().getPhysical() < outdatedLeaseThreshold` is crucial there. But, may be I should add the pendings check in  `chooseCandidateAndCreateNewLease` too (not instead)?

Reply:
And what about the case of creating the new lease?
Probably, the both cases where `writeNewLease` is used should be covered. Maybe you should combine this logic with `nextLeaseHolder` method?

Reply:
Done